### ISSUE TITLE
Maxwell can crash on restart if binlog disconnection causes connection to different db host

### ISSUE THREAD

**Issue by tamimkh** [zendesk/maxwell#1863]
It seems like when maxwell is [disconnected from a host](https://github.com/zendesk/maxwell/blob/master/src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java#L762) it attempts to [re-connect](https://github.com/zendesk/maxwell/blob/master/src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java#L494) to the [host](https://github.com/zendesk/maxwell/blob/master/src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java#L505-L506) but it does no checks to determine if the db host it just re-connected to is the same db host as before. 

This has the potential to cause issues if maxwell can connect to multiple db hosts (ie: a DNS record re-presents the hosts maxwell can connect to or maxwell connects via a load balancer) and the cluster maxwell is connecting to does not use GTIDs. The most severe of these issues is if this re-connection succeeds without problems (ie: the binlog positions between the initial server (`DB1`) and server it failed over to (`DB2`) are "similar") and maxwell starts writing positions for `DB2` without realizing they are for `DB1`. If the positions drift over time the state of the maxwell positions table can make it impossible for maxwell to re-connect back to the `DB1`.  If maxwell is restarted after this sequence of events it will immediately crash since the positions it will try to use to connect to `DB1` are not valid and point to a random place in a binlog. 

Here is a diagram that explains the sequence of events: 
```mermaid
sequenceDiagram
    Maxwell->>+DB1: Binlog Connected
    Maxwell->>DB1: Binlog File N1, Binlog Pos X1
    Maxwell->>DB1: Network blip, Binlog Server Disconnected
    Maxwell->>+DB2: Reconnect, but DB2 (Doesn't check Server ID)
    Maxwell->>+DB2: Continues Binlog File N1, Binlog Pos X1
    Maxwell->>+DB2: Continues Binlog File N1, Binlog Pos X2
    Maxwell->>+DB2: Continues Binlog File N1, Binlog Pos X3
    Maxwell->>DB2: Network blip, Binlog Server Disconnected
    Maxwell->>+DB1: Binlog re-connected (Doesn't check Server ID)
    Maxwell->>+DB1: Tries Binlog File N1, Binlog Pos X3
    Maxwell->>+DB1: Crash Due to Invalid Binlog Pos
    loop Crash Loop
      Maxwell->>+DB1: Crash on restart due invalid stored position in positions table
end 
```

It seems like this can be fixed by either attempting recovery upon detecting that the server has changed on re-connection or by not attempting re-connection at all and instead restarting maxwell on disconnects, maybe under a configuration flag. We would be happy to contribute a PR to maxwell to try to fix this issue but would like thoughts on what the best fix for this issue would be

**Comment by osheroff** [comment]
sure, would happily accept a patch to deal with this corner case.  

I think the trick here is going to be to add some code in mysql-binlog-connector to verify that the serverID hasn't changed.  so like upon reconnect you'd call:

```
  if ( !guid_mode) { 
     client.connect(5000, old_master_server_id);
  }
```

this would then use the code here:

https://github.com/osheroff/mysql-binlog-connector-java/blob/c19b1add648232dbd77151e40555927997a42ff5/src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java#L720

to verify that the server id hadn't changed upon reconnect.  If the server id changes, it throws an exception and crashes the process.

make sense?

**Comment by tamimkh** [comment]
Yup it does. 

However, if maxwell crashing is the best way to go about this I am wondering if modifying [this](https://github.com/zendesk/maxwell/blob/master/src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java#L494) to be something like the following would work as well:
```
Long oldMasterId = client.getMasterServerId();
tryReconnect();
if (client.getMasterServerId() != oldMasterId) {
	throw new Exception(
		"Master id changed from " + oldMasterId + "to " + client.getMasterServerId() + " while using binlog coordinate positioning. Cannot continue with the info that we have");
}
```

Or do you feel like this is something should be part of the `mysql-binlog-connector-java` library and as a result the checking for this change and throwing of the exception should happen in there?

**Comment by osheroff** [comment]
i think it'd be fine to do it how you're suggesting too.  I could definitely see it being useful to other users of binlog-connector, but it doesn't bother me if it's maxwell only either.

**Comment by tamimkh** [comment]
@osheroff Do you know if it would be possible to get a release of maxwell that includes https://github.com/zendesk/maxwell/pull/1866? Unfortunately, it has impacted us in the past and requires manual intervention in order to get maxwell back into a working state after it occurs.

**Comment by osheroff** [comment]
released in 1.37.6
